cmd/checkpoint/main.go

```diff
@@ -0,0 +1,157 @@
+package main
+
+import (
+	"bytes"
+	"encoding/json"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"k8s.io/kubernetes/pkg/api/v1"
+)
+
+var manifestBytes = []byte(`
+{
+"apiVersion": "v1",
+"kind": "Pod",
+"metadata": {
+  "name": "temp-apiserver",
+  "namespace": "kube-system"
+},
+"spec": {}
+}
+`)
+
+var (
+	tempAPIServer = []byte("temp-apiserver")
+	kubeAPIServer = []byte("kube-apiserver")
+)
+
+const (
+	kubeletAPIPodsURL = "http://localhost:10255/pods"
+	ignorePath        = "/etc/kubernetes/_manifests/apiserver.json"
+	activePath        = "/etc/kubernetes/manifests/apiserver.json"
+)
+
+func main() {
+	log.Println("begin apiserver checkpointing...")
+	// Create ignore dir.
+	if err := os.MkdirAll(filepath.Dir(ignorePath), 0755); err != nil {
+		log.Fatal(err)
+	}
+	run()
+}
+
+func run() {
+	var manifest v1.Pod
+	if err := json.Unmarshal(manifestBytes, &manifest); err != nil {
+		log.Fatal(err)
+	}
+	for {
+		rawPods := getPods()
+		switch {
+		case bothAPIServersRunning(rawPods):
+			log.Println("both temp and kube apiserver running, removing temp apiserver")
+			// Both the self-hosted API Server and the temp API Server are running.
+			// Remove the temp API Server manifest from the config dir so that the
+			// kubelet will stop it.
+			if err := os.Remove(activePath); err != nil {
+				log.Println(err)
+			}
+		case kubeSystemAPIServerRunning(rawPods):
+			log.Println("kube-apiserver found, creating temp-apiserver manifest")
+			// The self-hosted API Server is running. Let's snapshot the pod,
+			// clean it up a bit, and then save it to the ignore path for
+			// later use.
+			var podList v1.PodList
+			var apiPod v1.Pod
+			if err := json.Unmarshal(rawPods, &podList); err != nil {
+				log.Fatal(err)
+			}
+			for _, p := range podList.Items {
+				if strings.Contains(p.Name, string(kubeAPIServer)) {
+					apiPod = p
+					break
+				}
+			}
+			cleanVolumes(&apiPod)
+			modifyInsecurePort(&apiPod)
+			manifest.Spec = apiPod.Spec
+			m, err := json.Marshal(manifest)
+			if err != nil {
+				log.Fatal(err)
+			}
+			if err := ioutil.WriteFile(ignorePath, m, 0644); err != nil {
+				log.Fatal(err)
+			}
+
+		default:
+			log.Println("no apiserver running, installing temp apiserver static manifest")
+			b, err := ioutil.ReadFile(ignorePath)
+			if err != nil {
+				log.Println(err)
+				continue
+			}
+			if err := ioutil.WriteFile(activePath, b, 0644); err != nil {
+				log.Println(err)
+			}
+		}
+		time.Sleep(time.Second)
```

Comment:
The reason I'm doing it rather frequently is to stay up to date. Consider if the apiserver pod is changed, we don't get notified, so instead we checkpoint frequently to ensure we're mostly up to date. I'm fine with backing this off a bit, or maybe diffing the manifest against the on-disk manifest, but that seems more complicated for little gain.